### DISCUSSION TITLE
SWC is installed on production installation

### DISCUSSION THREAD

**Discussion by felixmosh** [vercel/next.js#33398] [Help]
### Run `next info` (available from version 12.0.8 and up)

    Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 21.2.0: Sun Nov 28 20:28:54 PST 2021; root:xnu-8019.61.5~1/RELEASE_X86_64
    Binaries:
      Node: 14.18.1
      npm: 6.14.15
      Yarn: 1.19.1
      pnpm: N/A
    Relevant packages:
      next: 12.0.8
      react: 17.0.2
      react-dom: 17.0.2

### What version of Next.js are you using?

12.0.8

### What version of Node.js are you using?

14.18.1

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

other

### Describe the Bug

When installing a production install using `yarn install --frozen-lockfile --production`

The size of Next.js is 77M, 
```
3.2M	node_modules/caniuse-lite
 33M	node_modules/next
 44M	node_modules/@next
```

when `@next` size is basically due to `swc`.
```
 24K	node_modules/@next/env
 60K	node_modules/@next/react-refresh-utils
 44M	node_modules/@next/swc-darwin-x64
```

Why it got installed in production installation?

### Expected Behavior

`swc` should not be installed in production installation.

### To Reproduce

make an empty folder with the following `package.json`:
```json
{
  "name": "next-swc-install",
  "version": "1.0.0",
  "main": "index.js",
  "author": "felixmosh",
  "license": "MIT",
  "dependencies": {
    "next": "^12.0.8"
  }
}
```

And run `yarn install --frozen-lockfile --production`

**Comment by timneutkens** [comment]
Because npm does not give us any way to mark "development" vs "production", so there's no differentiation between what should be needed in production vs development. 

If you want the Next.js app to run without node_modules and only the deps needed for the app you can use the new experimental flag `outputStandalone` that leverages the node-file-trace files to create a standalone Next.js app: https://nextjs.org/docs/advanced-features/output-file-tracing#automatically-copying-traced-files-experimental.

**Reply by timneutkens** [reply]
Converted to discussions as the proposed behavior is not possible because of upstream ecosystem limitations (npm) and we have an experimental flag implemented for standalone outputs.

**Reply by felixmosh** [reply]
What about `devDependencies`? `--production` flag meant for this specific thing, it won't install devDependencies.

**Reply by timneutkens** [reply]
`devDependencies` of a package you're installing are never installed. Either way even if that was possible the new standalone output is much better as it only includes the deps that are actually used by the app, so it's not even the `next` package specifically but also all others.

**Reply by felixmosh** [reply]
Then, why swc is not in `devDeps`?

**Reply by timneutkens** [reply]
> `devDependencies` of a package you're installing are never installed

`npm install next` will not install the `next` package `devDependencies`. We are not able to say "install this in dev" "install this when `npm install --production`" because npm does not support a way to flag deps like that.

**Reply by felixmosh** [reply]
I've investigated it, and `@next/swc-darwin-x64` was installed since it appears in `optionalDependencies` of `next`.

```json
"optionalDependencies": {
    "@next/swc-android-arm64": "12.0.8",
    "@next/swc-darwin-arm64": "12.0.8",
    "@next/swc-darwin-x64": "12.0.8",
    "@next/swc-linux-arm-gnueabihf": "12.0.8",
    "@next/swc-linux-arm64-gnu": "12.0.8",
    "@next/swc-linux-arm64-musl": "12.0.8",
    "@next/swc-linux-x64-gnu": "12.0.8",
    "@next/swc-linux-x64-musl": "12.0.8",
    "@next/swc-win32-arm64-msvc": "12.0.8",
    "@next/swc-win32-ia32-msvc": "12.0.8",
    "@next/swc-win32-x64-msvc": "12.0.8"
  },
```

**Reply by felixmosh** [reply]
I've found that adding `--ignore-optional` to the install command helps.

**Reply by felixmosh** [reply]
Is `@next/swc-darwin-x64` required for production run?

**Reply by MauriceArikoglu** [reply]
@timneutkens @leerob Is `@next/swc-darwin-x64` required for production run?

**Reply by mostlylikeable** [reply]
Would be nice to understand if swc should be required in standalone production builds. I manually `rm`'d in production image build and got an error like this (next: v14.2.2):
```
code: 'MODULE_NOT_FOUND',
  at Module.load (node:internal/modules/cjs/loader:1203:32) {
  at Module._extensions..js (node:internal/modules/cjs/loader:1422:10)
  at Module._compile (node:internal/modules/cjs/loader:1364:14)
  at Object.<anonymous> (/app/node_modules/next/dist/shared/lib/constants.js:283:34)
  at require (node:internal/modules/helpers:177:18)
  at mod.require (/app/node_modules/next/dist/server/require-hook.js:65:28)
  at Module.require (node:internal/modules/cjs/loader:1231:19)
  at Module._load (node:internal/modules/cjs/loader:981:27)
  at /app/node_modules/next/dist/server/require-hook.js:55:36
  at Module._resolveFilename (node:internal/modules/cjs/loader:1140:15)
    - /app/server.js
    - /app/node_modules/next/dist/server/next.js
    - /app/node_modules/next/dist/server/config.js
    - /app/node_modules/next/dist/shared/lib/constants.js
  Require stack:
    Error: Cannot find module '@swc/helpers/_/_interop_require_default'
```

I assumed since it was part of compilation that it would be ok to remove for standalone app, but this error seems to indicate otherwise.

**Reply by ardokirsipuu** [reply]
Those swc helpers are needed. I just removed the @swc/core-&lt;yourplatform-arch-etc&gt; folder since that's the one that takes considerably more space than the rest of it.
![image](https://github.com/user-attachments/assets/36910d5b-380e-4198-a98d-514e4e01d81d)
I'm using the standalone output so I just added this to my next.config.js file (see [docs](https://nextjs.org/docs/pages/api-reference/config/next-config-js/output#caveats)):
```
  output: "standalone",
  outputFileTracingExcludes: {
    "./**/*": [
      "node_modules/@swc/core-*",
    ],
  },
```
and then the large swc binary will not end up in the standalone/node_modules folder anymore.
